### DISCUSSION TITLE
Can't find a way by use alert_text_args

### DISCUSSION THREAD

**Discussion by SamMc75** [jertel/elastalert2#945] [Q&A] [answered]
Hi!
Can anyone help me.

If I use my rule without use alert_text_type, alert_text_args and alert_text I receive this alert for my rule:

At least 1 events occurred between 2022-09-14 13:31 and 2022-09-14 13:36

@timestamp: 2022-09-14T13:36:41.049Z
@version: 1
_id: PHE3PIMBy0RiEk67TnW_
_index: gpbapi-app-log-2022.09.14
_type: _doc
host: 131.107.1.35
level: INFO
level_value: 20000
logger_name: ru.controller.ApiController
message: {"currentDateTime":"2022-09-14T16:36:41.048134","events":[{"eventId":"","msisdn":"","caseId":"","occurDateTime":"1970-01-01T00:00:00","client":null,"data":[]}]}
num_hits: 80
num_matches: 1
port: 46558
thread_name: https-jsse-nio-0.0.0.0-9001-exec-15

It's okay.

But if I want to use alert_text_type, alert_text_args and alert_text and write to rule like:
At least {{num_hits}} events occurred between {{starttime}} and {{endtime}} .

I receive text in alert:
At least 80 events occurred between 2022-09-14 13:31 and  . 

So instead value of endtime I receive blank.
Why?

If I write a rule:
At least {{num_hits}} events occurred. CurrentDateTime is {{message.currentDateTime}} .

I receive text in alert:
At least 80 events occurred. CurrentDateTime is  . So instead {{message.currentDateTime}} I receive blank value.

I don't understand how I can use alert_text_args.

**Comment by jertel** [comment, marked as answer]
Did you see this similar question in the FAQ? 

https://elastalert2.readthedocs.io/en/latest/recipes/faq.html?highlight=alert_text_args#how-can-i-change-what-s-in-the-alert

**Reply by SamMc75** [reply]
Hi, I saw it, by I don't find answer about use starttime and endtime. Why starttime is present in alert but not endtime?
For currentDateTime from message value: I see that I'm using it wrong. But how I can use value of field currentDateTime via alert_text_args?

**Reply by jertel** [reply]
`starttime` is available within the match, but `endtime` is not. In the default alert text, the endtime is actually the same value as the `@timestamp` of the event. So you could mimic that similar behavior by referencing `@timestamp` instead of endtime.

To access nested subfields via alert_text_args, seperate the subfield from the parent field with a period. 

The below rule example illustrates both of the above statements, using the Kibana eCommerce Sample Data:
```
name: clothing_order_place46
is_enabled: true
type: any

filter:
   - term:
         currency: EUR

index: "*"

timestamp_field: order_date

alert_text_type: alert_text_only
alert_text: |-
  Women's Clothing order placed.
  Details of the event:
          - currency: {0}
          - email: {1}
          - products.manufacturer: {2}
          - geoip.continent_name: {3}
          - taxful_total_price: {4}
          - Search Start Time: {5} 
          - Search End Time: {6}

alert_text_args:
  - currency
  - email
  - products[0]manufacturer
  - geoip.continent_name
  - taxful_total_price
  - starttime
  - order_date
```

And the resulting Slack alert:
![image](https://user-images.githubusercontent.com/16639611/190216580-1e124f0f-b7ef-437f-87a1-0cd43795ef87.png)

**Reply by SamMc75** [reply]
Okay, thanks a lot for answer about endtime. 
I see that in rule for example time format both Start time and End time is different and not same. How I can in alert_text result time in same format? 
And how I can result format to UTC + my region?
I tried use some tunes in rule file yaml (timestamp_format and use_local_time) but it's not work.

**Reply by jertel** [reply]
This is a similar discussion and may help: https://github.com/jertel/elastalert2/discussions/342

**Reply by SamMc75** [reply]
Thanks for reply. But my problem is message value is string. I think I've convert this string to JSON like {{ message | tojson(indent=2) }}  and then I'll try to work with them as JSON , but I'm not sure that right a way in my case

**Reply by jertel** [reply]
The correct way to handle the message field is with a filebeat or logstash pipeline that converts the JSON into actual indexed fields stored in the Elastic index.

**Reply by SamMc75** [reply]
Yes, it will be great but unfortunately it's not possible because field "message" as is use many apps.

**Reply by jertel** [reply]
Yes, message is a commonly used field for generic messages. But that doesn't mean you cannot have a pipeline decode it as JSON, and if successful, store the decoded fields JSON. Pipelines are very capable of handling disparate types of data and selectively processing specific pieces, and formats.
